x/gamm/pool-models/balancer/amm.go

```diff
@@ -131,7 +131,11 @@ func (p *Pool) SwapInAmtGivenOut(
 	if err != nil {
 		return sdk.Coin{}, sdkerrors.Wrapf(types.ErrInvalidMathApprox, "token amount is zero or negative")
 	}
-	tokenInCoin, _ := tokenInDecCoin.TruncateDecimal()
+	tokenInCoin, tokenInDecimal := tokenInDecCoin.TruncateDecimal()
+	// if tokenInDecimal is non-zero, we add 1 to the tokenInCoin
+	if tokenInDecimal.Amount.IsPositive() {
+		tokenInCoin.Amount.AddRaw(1)
```

Comment:
don't we have to do `tokenInCoin = tokenInCoin.Amount.AddRaw(1)` ?

Reply:
Also, I'm curious why we add 1 in the case of non-zeros!

Reply:
ah yeah your right, good catch.

The idea here is that I want to get exactly `100` tokens out after the trade. But to do that, I have to provide say `50.82` tokens as input. Tokens are stored as integers, so before we were only requiring `50` tokens to be sent. But actually it should have been `51` when its converted to integers.

Reply:
I'm going to revert this change for Balancer, and fix it in a second PR that fixes CalcSpotPrice to be an integer API